mojo/core/broker_castanets.cc

```diff
@@ -268,6 +287,51 @@ void BrokerCastanets::SyncSharedBufferImpl(const base::UnguessableToken& guid,
   node_channel_->AddSyncFence(guid, fence_id, write_lock);
 }
 
+void BrokerCastanets::SyncSharedBufferImpl2d(const base::UnguessableToken& guid,
+                                             uint8_t* memory,
+                                             size_t offset,
+                                             size_t sync_size,
+                                             size_t mapped_size,
+                                             size_t width,
+                                             size_t stride,
+                                             bool write_lock) {
+  CHECK_GE(mapped_size, offset + sync_size);
+  BufferSyncData* buffer_sync = nullptr;
+  void* extra_data = nullptr;
+  Channel::MessagePtr out_message =
+      CreateBrokerMessage(BrokerMessageType::BUFFER_SYNC_2D, 0, sync_size,
+                          &buffer_sync, &extra_data);
+
+  FenceId fence_id = GenerateRandomFenceId();
+  buffer_sync->guid_high = guid.GetHighForSerialization();
+  buffer_sync->guid_low = guid.GetLowForSerialization();
+  buffer_sync->fence_id = fence_id;
+  buffer_sync->offset = offset;
+  buffer_sync->sync_bytes = sync_size;
+  buffer_sync->buffer_bytes = mapped_size;
+  buffer_sync->width = width;
+  buffer_sync->stride = stride;
+
+  size_t put_bytes = 0;
+  size_t put_offset = offset;
+
+  while (put_bytes != sync_size) {
+    memcpy(static_cast<uint8_t*>(extra_data) + put_bytes, memory + put_offset,
+           width);
+    put_offset += stride;
+    put_bytes += width;
+  }
+
```

Comment:
base::AutoLock lock(sync_lock_); is necessary here.

Reply:
Done. Thanks.